src/lib/operators/insert.cpp

```diff
@@ -107,136 +93,132 @@ std::shared_ptr<const Table> Insert::_on_execute(std::shared_ptr<TransactionCont
-      if (source_chunk_depleted) {
-        source_chunk_id++;
-        source_chunk_start_index = 0u;
+
+      if (num_rows_current_iteration == source_chunk_remaining_rows) {
+        // Proceed to next source Chunk
+        ++source_row_id.chunk_id;
+        source_row_id.chunk_offset = 0;
+      } else {
+        source_row_id.chunk_offset += num_rows_current_iteration;
       }
-    }
 
-    for (auto i = start_index; i < start_index + current_num_rows_to_insert; i++) {
-      // we do not need to check whether other operators have locked the rows, we have just created them
-      // and they are not visible for other operators.
-      // the transaction IDs are set here and not during the resize, because
-      // tbb::concurrent_vector::grow_to_at_least(n, t)" does not work with atomics, since their copy constructor is
-      // deleted.
-      target_chunk->get_scoped_mvcc_data_lock()->tids[i] = context->transaction_id();
-      _inserted_rows.emplace_back(RowID{target_chunk_id, i});
+      target_chunk_offset += num_rows_current_iteration;
+      target_chunk_range_remaining_rows -= num_rows_current_iteration;
     }
-
-    input_offset += current_num_rows_to_insert;
-    start_index = 0u;
   }
 
   return nullptr;
 }
 
 void Insert::_on_commit_records(const CommitID cid) {
-  for (auto row_id : _inserted_rows) {
-    auto chunk = _target_table->get_chunk(row_id.chunk_id);
-
-    auto mvcc_data = chunk->get_scoped_mvcc_data_lock();
-    mvcc_data->begin_cids[row_id.chunk_offset] = cid;
-    mvcc_data->tids[row_id.chunk_offset] = 0u;
+  for (const auto& target_chunk_range : _target_chunk_ranges) {
+    const auto target_chunk = _target_table->get_chunk(target_chunk_range.chunk_id);
+    auto mvcc_data = target_chunk->get_scoped_mvcc_data_lock();
+
+    for (auto chunk_offset = target_chunk_range.begin_chunk_offset; chunk_offset < target_chunk_range.end_chunk_offset;
+         ++chunk_offset) {
+      mvcc_data->begin_cids[chunk_offset] = cid;
+      mvcc_data->tids[chunk_offset] = 0u;
+    }
   }
 }
 
 void Insert::_on_rollback_records() {
-  for (auto row_id : _inserted_rows) {
-    auto chunk = _target_table->get_chunk(row_id.chunk_id);
-    // We set the begin and end cids to 0 (effectively making it invisible for everyone) so that the ChunkCompression
-    // does not think that this row is still incomplete. We need to make sure that the end is written before the begin.
-    chunk->get_scoped_mvcc_data_lock()->end_cids[row_id.chunk_offset] = 0u;
-    std::atomic_thread_fence(std::memory_order_release);
-    chunk->get_scoped_mvcc_data_lock()->begin_cids[row_id.chunk_offset] = 0u;
-
-    chunk->get_scoped_mvcc_data_lock()->tids[row_id.chunk_offset] = 0u;
+  for (const auto& target_chunk_range : _target_chunk_ranges) {
+    const auto target_chunk = _target_table->get_chunk(target_chunk_range.chunk_id);
+    auto mvcc_data = target_chunk->get_scoped_mvcc_data_lock();
```

Comment:
Do we need to hold the lock for the entire inner loop?

Reply:
Apparently we do, otherwise someone might call "shrink" on the mvcc data while we access it. Acquiring the lock for each row individually seems bad as well.